dask/dataframe/multi.py

```diff
@@ -419,6 +420,233 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
                          indicator=indicator)
 
 
+###############################################################
+# ASOF Join
+###############################################################
+
+def fix_overlap(ddf):
+    """ Ensures that ddf.divisions are all distinct and that the upper bound on
+    each partition is exclusive
+    """
+    if not ddf.known_divisions:
+        raise ValueError("Can only fix overlap when divisions are known")
+
+    def body(df, index):
+        return df.drop(index, inplace=True) if index in df else df
+
+    def overlap(df, index):
+        return df.loc[[index]] if index in df else None
+
+    dsk = dict()
+    name = 'fix-overlap-' + tokenize(ddf)
+
+    n = len(ddf.divisions) - 1
+    divisions = []
+    for i in range(n):
+        if i > 0 and ddf.divisions[i - 1] == ddf.divisions[i]:
+            frames = dsk[(name, len(divisions) - 1)][1]
+        else:
+            frames = []
+            if i > 0:
+                frames.append((overlap, (ddf._name, i - 1), ddf.divisions[i]))
+            divisions.append(ddf.divisions[i])
+            dsk[(name, len(divisions) - 1)] = (pd.concat, frames)
+
+        if i == n - 1 or ddf.divisions[i + 1] == ddf.divisions[i]:
+            frames.append((ddf._name, i))
+        else:
+            frames.append((body, (ddf._name, i), ddf.divisions[i + 1]))
+    divisions.append(ddf.divisions[-1])
+
+    graph = HighLevelGraph.from_collections(name, dsk, dependencies=[ddf])
+    return new_dd_object(graph, name, ddf._meta, divisions)
+
+
+def compute_tails(ddf, by=None):
+    """ For each partition, returns the last row of the most recent nonempty
+    partition.
+    """
+    empty = ddf._meta.iloc[0:0]
+
+    def none_to_empty(df):
+        return empty if df is None else df
+
+    if by is None:
+        def last_tail(L, R):
+            if R is None or R.empty:
+                return L
+            return R.tail(1)
+
+        return prefix_reduction(last_tail, ddf, postprocess=none_to_empty)
```

Comment:
Passing dynamically generated functions into dask graphs is a bit suboptimal, particularly when using a distributed cluster.  It's significantly more expensive to serialize a dynamically generated function in Python than one that is defined within a module.

Reply:
To be more clear, can I ask you to not define functions within other functions, if they are going to be passed into a task graph.  Functions like `last_tail` or `none_to_empty` defined just above should be moved outside of the `compute_tails` definition if they are going to end up in the task graph (which I think that they are).  This happens in a few other places as well.

What you've done works, but is slightly less efficient for reasons that are probably not so important to get into.

Reply:
I’m on it (but I’m away from my laptop until Monday, sorry).